runtime/runtime/src/actions.rs

```diff
@@ -473,9 +476,32 @@ pub(crate) fn action_implicit_account_creation_transfer(
         }
         // Invariant: The `account_id` is implicit.
         // It holds because in the only calling site, we've checked the permissions before.
-        AccountType::EthImplicitAccount | AccountType::NamedAccount => {
-            panic!("must be near-implicit")
+        AccountType::EthImplicitAccount => {
+            if checked_feature!("stable", EthImplicit, current_protocol_version) {
+                // TODO(eth-implicit) Use real Wallet Contract.
+                let wallet_contract = wallet_contract_placeholder();
+                let storage_usage = fee_config.storage_usage_config.num_bytes_account
+                    + fee_config.storage_usage_config.num_extra_bytes_record
+                    + wallet_contract.code().len() as u64;
+
+                *account =
+                    Some(Account::new(transfer.deposit, 0, *wallet_contract.hash(), storage_usage));
+
+                set_code(state_update, account_id.clone(), &wallet_contract);
```

Comment:
I don't think we should literally deploy the wallet contract code. Doing so has the following drawbacks:

1. Makes upgrades to the wallet contract more complex in the future (we would need to migrate all instances of it in the state).
2. Charges the user for storing the contract (via storage staking), creating a barrier to entry for on-boarding EVM users (they need at least 2 or 3 $NEAR and those tokens are "lost" in the sense that they will be permanently locked up in storage staking).

I think it would be better if we could "conceptually deploy" the wallet contract to all Eth implicit accounts by storing some kind of reference value instead. The runtime would need to know when it sees such a reference as the code then it should look up the protocol-defined wallet contract code to execute (it would be statically included in the nearcore binary for example). Then we only need to charge users for the storage consumed by referencing the contract (which should be very small and fit in within the zero balance accounts storage limit).

Reply:
Sounds like you're requesting a new protocol feature, that can be applied to more than just eth accounts.
IIRC, making duplicate contracts (mostly) free for (some) users was discussed a year ago. Revisiting it again is fine by me, but it seems to be out of scope of this PR.

Maybe @akhi3030 has more context on how contract storage should be charged.

Reply:
I don't have enough context on this feature and what precisely we are trying to build here to offer a lot of constructive feedback.  

Based on the above two comments, I agree with nikurt that it seems like an additional potentially orthogonal feature is being requested here that allows multiple accounts to operate using the same contract code.  

This feature has been discussed extensively in the past.  In general, we did not reach a conclusion where it made sense to implement support for such a feature in the protocol and it seemed like alternative ways of handling the use case outside of protocol sufficed.

Reply:
There is this past discussion on sharing contract code: https://github.com/near/NEPs/issues/478.

Reply:
I know there has been discussion about shared contracts before and there is a lot of complexity behind it. I don't mean to introduce a general mechanism as part of this feature. I only want to share one contract which is defined by the protocol, rather than a user. To me this sounds like a much simpler ask, but please correct me if I am wrong. I posted a similar comment to this one in the Zulip thread as well.

Reply:
As discussed in https://near.zulipchat.com/#narrow/stream/295306-pagoda.2Fcontract-runtime/topic/Eth-wallet.20support, I don't see any obvious issues with a single shared contract.  I've asked one other person from the team to validate my understanding.

Reply:
@birchmd and @staffik How can we proceed from here while making sure we're making incremental progress? Would you two be fine with adding a todo here and merging this code as is and later replacing it with the new scheme when it's ready?

Reply:
Sure, I'm fine with leaving this as a todo in this PR. After we agree on the design with the contract runtime team then this can be fixed in a follow-up PR.

Please still address the usage of `PROTOCOL_VERSION` in runtime (comment below) before merging this PR.